src/frontend/src/lib/services/load-user-profile.services.ts

```diff
@@ -0,0 +1,54 @@
+import type { UserProfile } from '$declarations/backend/backend.did';
+import { createUserProfile, getUserProfile } from '$lib/api/backend.api';
+import { i18n } from '$lib/stores/i18n.store';
+import { userProfileStore } from '$lib/stores/settings.store';
+import { toastsError } from '$lib/stores/toasts.store';
+import { UserProfileNotFoundError } from '$lib/types/errors';
+import type { Identity } from '@dfinity/agent';
+import { isNullish } from '@dfinity/utils';
+import { get } from 'svelte/store';
+
+const getUncertifiedProfile = async ({
+	identity
+}: {
+	identity: Identity;
+}): Promise<UserProfile | undefined> => {
+	try {
+		return await getUserProfile({ identity, certified: false });
+	} catch (err) {
+		if (err instanceof UserProfileNotFoundError) {
+			return undefined;
+		}
+		throw err;
+	}
+};
+
+const loadCertifiedUserProfile = async ({ identity }: { identity: Identity }): Promise<void> => {
+	try {
+		const profile = await getUserProfile({ identity, certified: true });
+		userProfileStore.set({ key: 'user-profile', value: profile });
+	} catch (err) {
+		// We ignore the error because this is a background task.
+		console.error('Failed to load certified user profile.', err);
+	}
+};
+
+export const loadUserProfile = async ({ identity }: { identity: Identity }): Promise<void> => {
+	try {
+		let profile = await getUncertifiedProfile({ identity });
+		if (isNullish(profile)) {
+			profile = await createUserProfile({ identity });
+		} else {
+			// We don't wait for this to complete because we want a smooth UX
+			// the uncertified data is enough for the user to start using the app.
+			loadCertifiedUserProfile({ identity });
```

Comment:
Two things come to my mind:

1. If the profile does not need to be created, I would already set the profile in the store as uncertified and not block the user. Then load with an update, and if the update fails, log out the user with an error message.

Otherwise, each login might end up being slower by a few seconds.

Or do you plan to not load the user profile on sign-in but, only when required?

2. Do we want to cache the profile on the client side as we do for the ETH address? If the profile is cached in IndexedDB (idb), we skip the query, skip the check to create it, and just do an update call in the background, which also signs out the user if unsuccessful.

Reply:
> If the profile does not need to be created, I would already set the profile in the store as uncertified and not block the user. Then load with an update

This is what the code does, isn't it?

I asked prodsec about caching the profile.

Moreover, I need to change the current caching because it caches all user profiles with the same key and we should add the principal in the key so that different users can be cached in the same device. This will come in another PR.

Yet, I don't see why we should follow a different path if we have the cache data.

Isn't the unsafe but updated data better than waiting for an update call?

If there is cached data, it means that getting the profile will work (or fail in other ways than NotFound), and it will skip creating. No need to specify it in the code.

Reply:
> This is what the code does, isn't it?
> ....
> Isn't the unsafe but updated data better than waiting for an update call?

Right; Sorry, I missed your comment ("We don't wait for this to complete...") in previous review.

Nevertheless, should we perhaps move the `userProfileStore.set` call before `loadCertifiedUserProfile`? Even if it's 99.999999% unlikely, code wise, there is maybe a potential race condition where the profile could first be set by the `update` in the store and then overwritten by the `query`?